File: Top Interview 150/C++ Code/Graph General/207-course-schedule.cpp
 // topological sort code 
 // check cycle in Directed graph
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& v) {
        vector<vector<int>>graph(n,vector<int>());
        vector<int>indegree(n,0);
        for(int i=0;i<v.size();i++){
            graph[v[i][1]].push_back(v[i][0]);
            indegree[v[i][0]]++;
        }
        queue<int>q;
        for(int i=0;i<n;i++){
            if(indegree[i]==0) q.push(i);
        }
        int cnt=0;
        while(q.size()){
            int u=q.front();
            q.pop();
            for(auto v:graph[u]){
                indegree[v]--;
                if(indegree[v]==0) q.push(v);
            }
            cnt++;
        }
        return cnt==n;
        
    }
};

File: Top Interview 150/Java Code/Graph General/207-course-schedule.java
 // topological sort code 
 // check cycle in Directed graph
class Solution {
    public boolean canFinish(int n, int[][] v) {
        List<List<Integer>>graph=new ArrayList<>();
        for(int i=0;i<n;i++){
            graph.add(new ArrayList<>());
        }
        int [] indegree=new int[n];
        for(int []it:v){
            graph.get(it[1]).add(it[0]);
            indegree[it[0]]++;
        }
        int cnt=0;
        Queue<Integer>q=new LinkedList<>();
        for(int i=0;i<n;i++){
            if(indegree[i]==0) q.add(i);
        }
        while(q.size()>0){
            int u=q.poll();
            for(int k:graph.get(u)){
                indegree[k]--;
                if(indegree[k]==0) q.add(k);
            }
            cnt++;
        }
        
        return cnt==n;
    }
}
